File: src/PDB.c
#include <stdio.h>
#include <stdlib.h>
#define SWAP(a,b) do{t=(a);(a)=(b);(b)=t;}while(0)

int _mr_rank1(int n, int *vec, int *inv)
{
    int s, t;
    if (n < 2) return 0;
 
    s = vec[n-1];
    SWAP(vec[n-1], vec[inv[n-1]]);
    SWAP(inv[s], inv[n-1]);
    return s + n * _mr_rank1(n-1, vec, inv);
}

long int get_rank(int n, byte *vec)
{
    int i, r, *v, *inv;
 
    v = malloc(n * sizeof(int));
    inv = malloc(n * sizeof(int));
 
    for (i = 0; i < n; ++i) {
        v[i] = vec[i];
        inv[vec[i]] = i;
    }
    r = _mr_rank1(n, v, inv);
    free(inv);
    free(v);
    return r;
}

int calcular_manhattan(int val1, int val2, int val3, int val4, int pos1, int pos2, int pos3, int pos4)
{
	int val = 0;

	int x = (val1) % 4;
	int y = (val1) / 4;
	int xs = (pos1) % 4;
	int ys = (pos1) / 4;
	val += abs(x - xs) + abs(y - ys);

	x = (val2) % 4;
	y = (val2) / 4;
	xs = (pos2) % 4;
	ys = (pos2) / 4;
	val += abs(x - xs) + abs(y - ys);

	x = (val3) % 4;
	y = (val3) / 4;
	xs = (pos3) % 4;
	ys = (pos3) / 4;
	val += abs(x - xs) + abs(y - ys);

	x = (val4) % 4;
	y = (val4) / 4;
	xs = (pos4) % 4;
	ys = (pos4) / 4;
	val += abs(x - xs) + abs(y - ys);

	return val;
}

int main(int argc, const char* argv[])
{
	int a, b, c, d;
	for (int i = 0; i < 16; ++i)
	{
		for (int j = 0; j < 16; ++j)
		{
			if (j != i)
				for (int k = 0; k < 16; ++k)
				{
					if (k != j && k != i)
					for (int l = 0; l < 16; ++l)
						{
							if (l != k && l != j && l != i)
							{
								a = calcular_manhattan(i, j, k, l, 0, 1, 4, 5);
								b = calcular_manhattan(i, j, k, l, 2, 3, 6, 7);
								c = calcular_manhattan(i, j, k, l, 8, 9, 12, 13);
								d = calcular_manhattan(i, j, k, l, 10, 11, 14, 15);
								printf("{%d, %d, %d, %d}, {%d, %d, %d, %d}\n", i, j, k, l, a, b, c, d);
							}
						}
				}
		}
	}
}

File: branch/pdb.py
def calcular_manhattan( val1,  val2,  val3,  val4,  pos1,  pos2,  pos3,  pos4):
	val = 0;

	if (val1 != 0):
		x = (val1) % 4;
		y = (val1) / 4;
		xs = (pos1) % 4;
		ys = (pos1) / 4;
		val += abs(x - xs) + abs(y - ys);

	if (val2 != 0):
		x = (val2) % 4;
		y = (val2) / 4;
		xs = (pos2) % 4;
		ys = (pos2) / 4;
		val += abs(x - xs) + abs(y - ys);

	if (val3 != 0):
		x = (val3) % 4;
		y = (val3) / 4;
		xs = (pos3) % 4;
		ys = (pos3) / 4;
		val += abs(x - xs) + abs(y - ys);
	
	if (val4 != 0):
		x = (val4) % 4;
		y = (val4) / 4;
		xs = (pos4) % 4;
		ys = (pos4) / 4;
		val += abs(x - xs) + abs(y - ys);

	return val;


r = [0,1,2,3,4,5,6,7,8,9,10,11,12,13,14,15]
m = [0,1,2,3]
val = [[[[[0 for x in m] for x in r] for x in r] for x in r] for x in r]
for i in r:
	for j in r:
		if i != j:
			for k in r:
				if k != j and k != i:
					for l in r:
						if l != k and l != j and l != i:
							val[i][j][k][l][0] = calcular_manhattan(i, j, k, l, 0, 1, 4, 5);
							val[i][j][k][l][1] = calcular_manhattan(i, j, k, l, 2, 3, 6, 7);
							val[i][j][k][l][2] = calcular_manhattan(i, j, k, l, 8, 9, 12, 13);
							val[i][j][k][l][3] = calcular_manhattan(i, j, k, l, 10, 11, 14, 15);
							#print "{"+str(i)+", "+str(j)+", "+str(k)+", "+str(l)+"}, {"+str(val[i][j][k][l][0])+", "+str(val[i][j][k][l][1])+", "+str(val[i][j][k][l][2])+", "+str(val[i][j][k][l][3])+"}"

print val
